crates/engine/tree/src/tree.rs

```diff
@@ -0,0 +1,126 @@
+use crate::{chain::PipelineAction, engine::DownloadRequest};
+use parking_lot::Mutex;
+use reth_beacon_consensus::{ForkchoiceStateTracker, InvalidHeaderCache, OnForkChoiceUpdated};
+use reth_engine_primitives::EngineTypes;
+use reth_primitives::SealedBlockWithSenders;
+use reth_rpc_types::{
+    engine::{CancunPayloadFields, ForkchoiceState, PayloadStatus},
+    ExecutionPayload,
+};
+use std::sync::Arc;
+
+/// Keeps track of the state of the tree.
+#[derive(Clone)]
+pub struct TreeState {
+    // TODO: this is shared state for the blocks etc.
+}
+
+impl TreeState {
+    fn buffer(&mut self) {}
+
+    fn insert_validated(&mut self) {}
+}
+
+/// Tracks the state of the engine api internals.
+///
+/// This type is shareable.
+#[derive(Clone)]
+pub struct EngineApiTreeState {
+    /// Tracks the received forkchoice state updates received by the CL.
+    forkchoice_state_tracker: ForkchoiceStateTracker,
+    /// Tracks the header of invalid payloads that were rejected by the engine because they're
+    /// invalid.
+    invalid_headers: Arc<Mutex<InvalidHeaderCache>>,
+
+    /// Tracks the state of the blockchain tree
+    tree_state: TreeState,
+}
+
+/// The type responsible for processing engine API requests.
+///
+/// TODO: design: should the engine handler functions also accept the response channel or return the
+/// result and the caller redirects the response
+pub trait EngineApiTreeHandler: Send + Sync + Clone {
```

Comment:
we also need a separate method for finalizing at block height

Reply:
and per discussion in discord we should also allow buffering & committing transactions separate from the state/tree/historical indices?

Reply:
these are tree implementation details that have nothing to do with the interfaces. the tree should be able to decide on its own when to buffer / commit